File: implementations/ruby/lib/provekit/claim_envelope.rb
# SPDX-License-Identifier: Apache-2.0
#
# Provekit::ClaimEnvelope -- claim-envelope minter (v1.2 layered shape).
#
# `mint_contract` / `mint_bridge` / `mint_implication` build a signed
# memento in the v1.2 LAYERED shape introduced by
# `protocol/specs/2026-05-03-substrate-layers-envelope-header-body.md`:
#
#   { "envelope": {...}, "header": {...}, "metadata": {...} }
#
#   * envelope = { signer, declaredAt, signature }
#       The signature is computed over JCS({"header": header,
#       "metadata": metadata}). The envelope's CID (= attestation CID)
#       is BLAKE3-512(JCS(envelope)) AFTER the signature has been
#       embedded.
#
#   * header   = substrate-load-bearing data the verifier reads:
#                schemaVersion, kind, cid, plus kind-specific REQUIRED
#                fields and the derived hashes (bindingHash,
#                propertyHash, verdict, inputCids) used by the
#                resolve/index pipeline.
#
#   * metadata = everything else (authoring attribution, lifecycle
#                strings like producedBy/producedAt, derived per-formula
#                hashes that are pure tooling convenience). Opaque to
#                the substrate verifier; signed transitively via the
#                envelope.
#
# Reference (authoritative):
#   implementations/rust/provekit-claim-envelope/src/lib.rs
#   implementations/python/provekit-lift-py-tests/src/provekit_lift_py_tests/claim_envelope.py
#
# Cross-kit byte-equivalence:
#   The Ruby output is byte-identical to the Rust + Python kits for the
#   same canonical input. See test/test_claim_envelope.rb.

require_relative "blake3"
require_relative "ir"
require_relative "signing"

module Provekit
  module ClaimEnvelope
    # The layered-shape schema version stamped into every memento header
    # emitted by this kit. Older flat mementos carry "1"; verifiers
    # branch on this string at load time.
    LAYERED_SCHEMA_VERSION = "2".freeze

    # ---- Errors ----------------------------------------------------------

    class Error < StandardError; end
    class EmptyContractError < Error; end
    class EmptyOutBindingError < Error; end

    # ---- Authoring (typed union mirrored from rust/python kits) ----------

    # Hand-authored by a human or kit, no inference.
    # `note` is optional; nil or empty string is treated as absent.
    AuthoringKitAuthor = Struct.new(:author, :note, keyword_init: true) do
      def initialize(author:, note: nil)
        super
      end
    end

    # Produced by a structural lifter from source.
    # `evidence` is the lifter's classification ("tests", "types", ...).
    # `source_cid` is optional; nil or empty string is treated as absent.
    AuthoringLift = Struct.new(:lifter, :evidence, :source_cid, keyword_init: true) do
      def initialize(lifter:, evidence:, source_cid: nil)
        super
      end
    end

    # Produced by an LLM, with prompt provenance.
    # `confidence` is a Float in [0, 1]; serialized as `(confidence *
    # 1000).to_i` (truncating toward zero for non-negative values; this
    # matches Rust's `(c * 1000.0) as i64` cast which truncates toward
    # zero, NOT rounding).
    # `rationale` is optional; nil or empty string is treated as absent.
    AuthoringLlm = Struct.new(
      :llm, :llm_version, :prompt_cid, :confidence, :rationale,
      keyword_init: true,
    ) do
      def initialize(llm:, llm_version:, prompt_cid:, confidence:, rationale: nil)
        super
      end
    end

    # ---- Result type -----------------------------------------------------

    # Result of minting a layered claim envelope. Three fields:
    # * canonical_bytes: JCS-canonical bytes of the full layered memento
    #   `{envelope, header, metadata}`. Binary String (ASCII-8BIT).
    # * cid: the attestation CID = BLAKE3-512(JCS(envelope)) AFTER the
    #   signature has been embedded. Self-identifying string form
    #   "blake3-512:<128 hex>".
    # * contract_cid: signer-independent content CID for contract
    #   mementos. Empty string for bridges and implications.
    Minted = Struct.new(:canonical_bytes, :cid, :contract_cid, keyword_init: true)

    # ---- Public surface --------------------------------------------------

    # Build a v1.2-layered ClaimEnvelope from a Provekit::IR::ContractDecl.
    #
    # Required:
    #   decl        -- ContractDecl with name, out_binding, and at least
    #                  one of pre/post/inv populated.
    #   signer      -- Provekit::Signing::Signer (Ed25519 seed + producer-id).
    #   produced_at -- ISO-8601 timestamp (ms precision, trailing 'Z').
    #
    # Optional:
    #   authoring   -- defaults to AuthoringKitAuthor.new(author: signer.producer_id).
    #   input_cids  -- defaults to []; sorted lex inside the header.
    def self.from_contract_decl(decl, signer, produced_at:, authoring: nil, input_cids: nil)
      authoring ||= AuthoringKitAuthor.new(author: signer.producer_id)
      mint_contract(
        contract_name: decl.name,
        out_binding: decl.out_binding,
        pre: decl.pre,
        post: decl.post,
        inv: decl.inv,
        produced_by: signer.producer_id,
        produced_at: produced_at,
        authoring: authoring,
        signer_seed: signer.seed,
        input_cids: input_cids || [],
      )
    end

    # Compute the **content** CID of a contract (signer-independent).
    #
    # Per `protocol/specs/2026-05-03-contract-cid-vs-attestation-cid.md`
    # §1, this is the BLAKE3-512 of the JCS encoding of the contract's
    # substrate-load-bearing fields: name, outBinding, and any of
    # pre/post/inv that are present. Two distinct signers attesting to
    # the same logical contract produce the same contractCid.
    def self.contract_cid(contract_name:, out_binding:, pre: nil, post: nil, inv: nil)
      obj = {
        "name" => contract_name,
        "outBinding" => out_binding,
      }
      obj["pre"]  = pre  unless pre.nil?
      obj["post"] = post unless post.nil?
      obj["inv"]  = inv  unless inv.nil?
      hash_value(obj)
    end

    # Mint a v1.2-layered contract claim envelope.
    # Byte-identical to Rust's `mint_contract` for the same canonical
    # inputs (see test/test_claim_envelope.rb cross-kit tests).
    def self.mint_contract(contract_name:, out_binding:, produced_by:, produced_at:,
                           authoring:, signer_seed:,
                           pre: nil, post: nil, inv: nil, input_cids: nil)
      if pre.nil? && post.nil? && inv.nil?
        raise EmptyContractError, "mint_contract: at least one of pre/post/inv must be present"
      end
      if out_binding.to_s.empty?
        raise EmptyOutBindingError, "mint_contract: outBinding must not be empty"
      end

      # DERIVED:
      #   propertyHash = hash(JCS({pre?, post?, inv?, outBinding}))
      #   bindingHash  = hash(JCS({producerId, contractName, propertyHash}))
      ph = {}
      ph["pre"]  = pre  unless pre.nil?
      ph["post"] = post unless post.nil?
      ph["inv"]  = inv  unless inv.nil?
      ph["outBinding"] = out_binding
      property_hash = hash_value(ph)

      binding_hash = hash_value({
        "producerId" => produced_by,
        "contractName" => contract_name,
        "propertyHash" => property_hash,
      })

      header_cid = contract_cid(
        contract_name: contract_name,
        out_binding: out_binding,
        pre: pre, post: post, inv: inv,
      )

      header = {
        "schemaVersion" => LAYERED_SCHEMA_VERSION,
        "kind" => "contract",
        "cid" => header_cid,
        "name" => contract_name,
        "outBinding" => out_binding,
      }
      header["pre"]  = pre  unless pre.nil?
      header["post"] = post unless post.nil?
      header["inv"]  = inv  unless inv.nil?
      header["verdict"] = "holds"
      header["bindingHash"] = binding_hash
      header["propertyHash"] = property_hash
      sorted_inputs = (input_cids || []).map(&:to_s).sort
      header["inputCids"] = sorted_inputs

      metadata = {
        "authoring" => authoring_to_value(authoring),
        "producedBy" => produced_by,
        "producedAt" => produced_at,
      }
      metadata["preHash"]  = hash_value(pre)  unless pre.nil?
      metadata["postHash"] = hash_value(post) unless post.nil?
      metadata["invHash"]  = hash_value(inv)  unless inv.nil?

      assemble_layered(header, metadata, produced_at, signer_seed, header_cid)
    end

    # Mint a v1.2-layered bridge claim envelope.
    # Byte-identical to Rust's `mint_bridge` for the same canonical inputs.
    #
    # DERIVED:
    #   bindingHash  = hash(JCS({sourceLayer, sourceSymbol}))
    #   propertyHash = hash("bridge:" + sourceSymbol)
    #   inputCids    = [target_contract_cid]
    def self.mint_bridge(produced_by:, produced_at:, source_symbol:, source_layer:,
                         target_contract_cid:, target_layer:, ir_arg_sorts:, ir_return_sort:,
                         signer_seed:, notes: "")
      arg_sorts = ir_arg_sorts.map(&:to_s)

      binding_hash = hash_value({
        "sourceLayer" => source_layer,
        "sourceSymbol" => source_symbol,
      })
      property_hash = hash_string("bridge:#{source_symbol}")

      header_cid = hash_value({
        "sourceSymbol" => source_symbol,
        "sourceLayer" => source_layer,
        "targetContractCid" => target_contract_cid,
        "targetLayer" => target_layer,
        "irArgSorts" => arg_sorts,
        "irReturnSort" => ir_return_sort,
      })

      header = {
        "schemaVersion" => LAYERED_SCHEMA_VERSION,
        "kind" => "bridge",
        "cid" => header_cid,
        "sourceSymbol" => source_symbol,
        "sourceLayer" => source_layer,
        "targetContractCid" => target_contract_cid,
        "targetLayer" => target_layer,
        "irArgSorts" => arg_sorts,
        "irReturnSort" => ir_return_sort,
        "verdict" => "holds",
        "bindingHash" => binding_hash,
        "propertyHash" => property_hash,
        "inputCids" => [target_contract_cid],
      }

      metadata = {
        "producedBy" => produced_by,
        "producedAt" => produced_at,
      }
      metadata["notes"] = notes unless notes.to_s.empty?

      assemble_layered(header, metadata, produced_at, signer_seed, "")
    end

    # Mint a v1.2-layered implication claim envelope.
    # Byte-identical to Rust's `mint_implication` for the same canonical
    # inputs.
    #
    # DERIVED:
    #   bindingHash  = hash(JCS({antecedentHash, consequentHash}))
    #   propertyHash = hash("implication:" + ah + ":" + ch)
    #   inputCids    = sorted([antecedent_cid, consequent_cid])
    def self.mint_implication(produced_by:, produced_at:, antecedent_hash:, consequent_hash:,
                              antecedent_cid:, consequent_cid:, antecedent_slot:, consequent_slot:,
                              prover:, prover_run_ms:, signer_seed:,
                              smt_lib_input: "", proof_witness: "")
      binding_hash = hash_value({
        "antecedentHash" => antecedent_hash,
        "consequentHash" => consequent_hash,
      })
      property_hash = hash_string("implication:#{antecedent_hash}:#{consequent_hash}")

      header_cid = hash_value({
        "antecedentHash" => antecedent_hash,
        "consequentHash" => consequent_hash,
        "antecedentCid" => antecedent_cid,
        "consequentCid" => consequent_cid,
        "antecedentSlot" => antecedent_slot,
        "consequentSlot" => consequent_slot,
      })

      input_cids_sorted = [antecedent_cid, consequent_cid].sort

      header = {
        "schemaVersion" => LAYERED_SCHEMA_VERSION,
        "kind" => "implication",
        "cid" => header_cid,
        "antecedentHash" => antecedent_hash,
        "consequentHash" => consequent_hash,
        "antecedentCid" => antecedent_cid,
        "consequentCid" => consequent_cid,
        "antecedentSlot" => antecedent_slot,
        "consequentSlot" => consequent_slot,
        "verdict" => "holds",
        "bindingHash" => binding_hash,
        "propertyHash" => property_hash,
        "inputCids" => input_cids_sorted,
      }

      metadata = {
        "producedBy" => produced_by,
        "producedAt" => produced_at,
        "prover" => prover,
        "proverRunMs" => prover_run_ms.to_i,
      }
      metadata["smtLibInput"]  = smt_lib_input  unless smt_lib_input.to_s.empty?
      metadata["proofWitness"] = proof_witness  unless proof_witness.to_s.empty?

      assemble_layered(header, metadata, produced_at, signer_seed, "")
    end

    # ---- Internals (mirror rust lib.rs / python claim_envelope.py) -------

    # BLAKE3-512(JCS(value)) -- "blake3-512:<128 hex>".
    def self.hash_value(value)
      Provekit::Blake3.hex(Provekit::IR::Jcs.encode(value))
    end

    # BLAKE3-512(s) -- s treated as UTF-8 bytes.
    def self.hash_string(s)
      Provekit::Blake3.hex(s.to_s)
    end

    # JCS-canonical bytes of `{"header": header, "metadata": metadata}`.
    # This is the message the envelope's Ed25519 signature covers
    # (substrate-layers spec §2 R2).
    def self.signing_bytes(header, metadata)
      Provekit::IR::Jcs.encode({"header" => header, "metadata" => metadata})
    end

    # Lower an Authoring block to a JCS-friendly Hash.
    # Mirrors `authoring_to_value` in rust kit / `_authoring_to_value` in
    # python kit. Empty notes/source_cid/rationale are treated as absent.
    def self.authoring_to_value(a)
      case a
      when AuthoringKitAuthor
        h = {"producerKind" => "kit-author", "author" => a.author}
        h["note"] = a.note if a.note && !a.note.empty?
        h
      when AuthoringLift
        h = {
          "producerKind" => "lift",
          "lifter" => a.lifter,
          "evidence" => a.evidence,
        }
        h["sourceCid"] = a.source_cid if a.source_cid && !a.source_cid.empty?
        h
      when AuthoringLlm
        # Rust's `(confidence * 1000.0) as i64` truncates toward zero.
        # Ruby's `Integer(x)` raises; `x.to_i` truncates toward zero
        # for both positive and negative finite floats, matching rust.
        confidence_i = (a.confidence * 1000.0).to_i
        h = {
          "producerKind" => "llm",
          "llm" => a.llm,
          "llmVersion" => a.llm_version,
          "promptCid" => a.prompt_cid,
          "confidence" => confidence_i,
        }
        h["rationale"] = a.rationale if a.rationale && !a.rationale.empty?
        h
      else
        raise ArgumentError, "unknown Authoring variant: #{a.class}"
      end
    end

    # Sign (header, metadata), embed signature into envelope, JCS the
    # full {envelope, header, metadata} triple. Returns Minted.
    #
    # Mirrors `assemble_layered` in rust:
    #   1. signer_str  = ed25519_pubkey_string(seed)
    #   2. signing_msg = JCS({"header": header, "metadata": metadata})
    #   3. signature   = ed25519_sign_string(seed, signing_msg)
    #   4. envelope    = {signer, declaredAt, signature}
    #   5. attestation_cid = BLAKE3-512(JCS(envelope))
    #   6. memento     = {envelope, header, metadata}
    #   7. canonical_bytes = JCS(memento) -> bytes
    def self.assemble_layered(header, metadata, declared_at, signer_seed, content_cid)
      signer_str = Provekit::Signing.ed25519_pubkey_string(signer_seed)
      signing_msg = signing_bytes(header, metadata)
      signature = Provekit::Signing.ed25519_sign_string(signer_seed, signing_msg)

      envelope = {
        "signer" => signer_str,
        "declaredAt" => declared_at,
        "signature" => signature,
      }
      envelope_jcs = Provekit::IR::Jcs.encode(envelope)
      attestation_cid = Provekit::Blake3.hex(envelope_jcs)

      memento_jcs = Provekit::IR::Jcs.encode({
        "envelope" => envelope,
        "header" => header,
        "metadata" => metadata,
      })

      Minted.new(
        canonical_bytes: memento_jcs.b,
        cid: attestation_cid,
        contract_cid: content_cid,
      )
    end
  end
end

File: implementations/java/provekit-self-contracts/src/main/java/com/provekit/claimenvelope/ClaimEnvelope.java
// SPDX-License-Identifier: Apache-2.0
//
// `mintContract` — build a signed claim envelope (the universal memento
// wrapper around a role-specific evidence body). Returns canonical bytes
// + CID.
//
// Java peer of implementations/rust/provekit-claim-envelope/src/lib.rs
// and implementations/csharp/Provekit.ClaimEnvelope/Mint.cs. v1.1.0
// hash widening: every hash is BLAKE3-512 (full 64-byte digest) prefixed
// with "blake3-512:". CIDs use the same form. NO truncation.
//
// Per-formula hashes (preHash/postHash/invHash) and propertyHash /
// bindingHash are DERIVED here from caller-supplied formula Values,
// never accepted from the caller. Validators recompute and reject
// mismatches.
//
// Scope vs. csharp peer: this Java port covers `mintContract`,
// `contractCid`, and `contractSetCid` — the load-bearing trio used by
// the rust orchestrator's `mint_from_ir_document` flow. Bridge and
// implication minting are tracked as follow-ups; they share the same
// {@link #mintInternal} substrate and can be added without disturbing
// the byte-equivalence already pinned here.

package com.provekit.claimenvelope;

import com.provekit.canonicalizer.Hash;
import com.provekit.canonicalizer.Jcs;
import com.provekit.canonicalizer.Value;
import com.provekit.proofenvelope.Sign;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public final class ClaimEnvelope {

    private ClaimEnvelope() {}

    // -----------------------------------------------------------------------
    // Schema CIDs (placeholders; mirrored byte-for-byte from rust/csharp)
    // -----------------------------------------------------------------------

    /** Placeholder schema CID for contract envelopes. */
    public static final String SCHEMA_CID_CONTRACT =
        "blake3-512:00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c01";

    // -----------------------------------------------------------------------
    // Result type
    // -----------------------------------------------------------------------

    public static final class MintedEnvelope {
        public final byte[] canonicalBytes;
        public final String cid;

        MintedEnvelope(byte[] canonicalBytes, String cid) {
            this.canonicalBytes = canonicalBytes;
            this.cid = cid;
        }
    }

    // -----------------------------------------------------------------------
    // mintContract input
    // -----------------------------------------------------------------------

    public static final class MintContractArgs {
        public final String contractName;
        public final Value pre; // nullable
        public final Value post; // nullable
        public final Value inv; // nullable
        public final String outBinding;
        public final String producedBy;
        public final String producedAt;
        public final List<String> inputCids;
        public final Authoring authoring;
        public final byte[] signerSeed;

        private MintContractArgs(Builder b) {
            this.contractName = b.contractName;
            this.pre = b.pre;
            this.post = b.post;
            this.inv = b.inv;
            this.outBinding = b.outBinding;
            this.producedBy = b.producedBy;
            this.producedAt = b.producedAt;
            this.inputCids = Collections.unmodifiableList(new ArrayList<>(b.inputCids));
            this.authoring = b.authoring;
            this.signerSeed = b.signerSeed.clone();
        }

        public static Builder builder() { return new Builder(); }

        public static final class Builder {
            private String contractName;
            private Value pre;
            private Value post;
            private Value inv;
            private String outBinding = "out";
            private String producedBy;
            private String producedAt;
            private List<String> inputCids = new ArrayList<>();
            private Authoring authoring;
            private byte[] signerSeed;

            public Builder contractName(String v) { this.contractName = v; return this; }
            public Builder pre(Value v) { this.pre = v; return this; }
            public Builder post(Value v) { this.post = v; return this; }
            public Builder inv(Value v) { this.inv = v; return this; }
            public Builder outBinding(String v) { this.outBinding = v; return this; }
            public Builder producedBy(String v) { this.producedBy = v; return this; }
            public Builder producedAt(String v) { this.producedAt = v; return this; }
            public Builder inputCids(List<String> v) {
                this.inputCids = new ArrayList<>(v);
                return this;
            }
            public Builder authoring(Authoring v) { this.authoring = v; return this; }
            public Builder signerSeed(byte[] v) {
                if (v == null || v.length != 32) {
                    throw new IllegalArgumentException("signer seed must be 32 bytes");
                }
                this.signerSeed = v.clone();
                return this;
            }

            public MintContractArgs build() {
                if (contractName == null) throw new IllegalStateException("contractName required");
                if (producedBy == null) throw new IllegalStateException("producedBy required");
                if (producedAt == null) throw new IllegalStateException("producedAt required");
                if (authoring == null) throw new IllegalStateException("authoring required");
                if (signerSeed == null) throw new IllegalStateException("signerSeed required");
                return new MintContractArgs(this);
            }
        }
    }

    // -----------------------------------------------------------------------
    // CIDs (signer-independent)
    // -----------------------------------------------------------------------

    /**
     * Compute the signer-independent {@code contractCid} for a contract.
     *
     * <p>Per spec 2026-05-03-contract-cid-vs-attestation-cid.md §1:
     * <pre>
     *   contractCid = blake3-512(JCS({name, outBinding, pre?, post?, inv?}))
     * </pre>
     * Two distinct signers attesting the same logical contract produce
     * the same contractCid. This is NOT the attestation CID (envelope
     * hash).
     */
    public static String contractCid(MintContractArgs args) {
        ArrayList<Map.Entry<String, Value>> entries = new ArrayList<>();
        entries.add(Map.entry("name", Value.ofString(args.contractName)));
        entries.add(Map.entry("outBinding", Value.ofString(args.outBinding)));
        if (args.pre != null) entries.add(Map.entry("pre", args.pre));
        if (args.post != null) entries.add(Map.entry("post", args.post));
        if (args.inv != null) entries.add(Map.entry("inv", args.inv));
        return hashValue(Value.ofObject(entries));
    }

    /**
     * Compute the {@code contractSetCid} from a list of signer-
     * independent contractCid strings.
     *
     * <p>Per spec 2026-05-03-contract-set-extension.md §1:
     * <pre>
     *   contractSetCid = blake3-512(JCS(&lt;sorted contractCids&gt;))
     * </pre>
     */
    public static String contractSetCid(Collection<String> contractCids) {
        ArrayList<String> sorted = new ArrayList<>(contractCids);
        // ASCII-sort for the contractCids (they are blake3-512:<hex> strings).
        sorted.sort(String::compareTo);
        ArrayList<Value> arr = new ArrayList<>(sorted.size());
        for (String s : sorted) arr.add(Value.ofString(s));
        return hashValue(Value.ofArray(arr));
    }

    // -----------------------------------------------------------------------
    // mintContract
    // -----------------------------------------------------------------------

    public static MintedEnvelope mintContract(MintContractArgs args) {
        if (args.pre == null && args.post == null && args.inv == null) {
            throw new IllegalArgumentException(
                "mintContract: at least one of pre/post/inv must be present");
        }
        if (args.outBinding == null || args.outBinding.isEmpty()) {
            throw new IllegalArgumentException("mintContract: outBinding must not be empty");
        }

        // Build evidence.body. Insertion order mirrors the C++/rust kits;
        // JCS sorts at emit time, so ordering is informational only.
        ArrayList<Map.Entry<String, Value>> body = new ArrayList<>();
        body.add(Map.entry("contractName", Value.ofString(args.contractName)));
        body.add(Map.entry("outBinding", Value.ofString(args.outBinding)));
        if (args.pre != null) {
            body.add(Map.entry("pre", args.pre));
            body.add(Map.entry("preHash", Value.ofString(hashValue(args.pre))));
        }
        if (args.post != null) {
            body.add(Map.entry("post", args.post));
            body.add(Map.entry("postHash", Value.ofString(hashValue(args.post))));
        }
        if (args.inv != null) {
            body.add(Map.entry("inv", args.inv));
            body.add(Map.entry("invHash", Value.ofString(hashValue(args.inv))));
        }
        body.add(Map.entry("authoring", args.authoring.toValue()));

        Value evidence = Value.ofObject(List.of(
            Map.entry("kind", Value.ofString("contract")),
            Map.entry("schema", Value.ofString(SCHEMA_CID_CONTRACT)),
            Map.entry("body", Value.ofObject(body))));

        // DERIVED:
        //   propertyHash = hash(canonical({pre?, post?, inv?, outBinding}))
        //   bindingHash  = hash(canonical({producerId, contractName, propertyHash}))
        ArrayList<Map.Entry<String, Value>> phEntries = new ArrayList<>();
        if (args.pre != null) phEntries.add(Map.entry("pre", args.pre));
        if (args.post != null) phEntries.add(Map.entry("post", args.post));
        if (args.inv != null) phEntries.add(Map.entry("inv", args.inv));
        phEntries.add(Map.entry("outBinding", Value.ofString(args.outBinding)));
        String propertyHash = hashValue(Value.ofObject(phEntries));

        Value bhObj = Value.ofObject(List.of(
            Map.entry("producerId", Value.ofString(args.producedBy)),
            Map.entry("contractName", Value.ofString(args.contractName)),
            Map.entry("propertyHash", Value.ofString(propertyHash))));
        String bindingHash = hashValue(bhObj);

        return mintInternal(
            bindingHash, propertyHash, "holds",
            args.producedBy, args.producedAt, args.inputCids,
            evidence, args.signerSeed);
    }

    // -----------------------------------------------------------------------
    // Internal: build wrapper, hash, sign, re-emit signed
    // -----------------------------------------------------------------------

    private static MintedEnvelope mintInternal(
            String bindingHash,
            String propertyHash,
            String verdict,
            String producedBy,
            String producedAt,
            List<String> inputCids,
            Value evidence,
            byte[] signerSeed) {

        // 1. Build the unsigned canonical envelope; hash it for the CID.
        Value unsignedV = buildEnvelopeForHashing(
            bindingHash, propertyHash, verdict, producedBy, producedAt, inputCids, evidence);
        byte[] unsignedBytes = Jcs.encodeUtf8(unsignedV);
        String cid = Hash.blake3_512(unsignedBytes);

        // 2. Sign the unsigned canonical bytes.
        String producerSig = Sign.ed25519SignString(signerSeed, unsignedBytes);

        // 3. Re-emit with cid + producerSignature appended; JCS re-sorts.
        ArrayList<Map.Entry<String, Value>> signedEntries =
            new ArrayList<>(unsignedV.asObject());
        signedEntries.add(Map.entry("cid", Value.ofString(cid)));
        signedEntries.add(Map.entry("producerSignature", Value.ofString(producerSig)));
        Value signedV = Value.ofObject(signedEntries);
        byte[] finalBytes = Jcs.encodeUtf8(signedV);
        return new MintedEnvelope(finalBytes, cid);
    }

    private static Value buildEnvelopeForHashing(
            String bindingHash,
            String propertyHash,
            String verdict,
            String producedBy,
            String producedAt,
            List<String> inputCids,
            Value evidence) {
        // Wrapper inputCids MUST be lex-sorted (spec §wrapper).
        ArrayList<String> sortedCids = new ArrayList<>(inputCids);
        sortedCids.sort(String::compareTo);
        ArrayList<Value> sortedArr = new ArrayList<>(sortedCids.size());
        for (String s : sortedCids) sortedArr.add(Value.ofString(s));

        return Value.ofObject(List.of(
            Map.entry("schemaVersion", Value.ofString("1")),
            Map.entry("bindingHash", Value.ofString(bindingHash)),
            Map.entry("propertyHash", Value.ofString(propertyHash)),
            Map.entry("verdict", Value.ofString(verdict)),
            Map.entry("producedBy", Value.ofString(producedBy)),
            Map.entry("producedAt", Value.ofString(producedAt)),
            Map.entry("inputCids", Value.ofArray(sortedArr)),
            Map.entry("evidence", evidence)));
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private static String hashValue(Value v) {
        return Hash.blake3_512(Jcs.encodeUtf8(v));
    }

    /** Unused but kept symmetric with the csharp peer; trivial helper. */
    @SuppressWarnings("unused")
    private static String hashString(String s) {
        return Hash.blake3_512(s.getBytes(StandardCharsets.UTF_8));
    }
}
